### PR TITLE
[ADVAPP-1064]: Refactor Task access to not use individual permissions to manage access

### PR DESCRIPTION
**Pull request by payal-canyon**
### Ticket(s) or GitHub Issue

- https://canyongbs.atlassian.net/browse/ADVAPP-1064

### Technical Description

> Refactor Task access to not use individual permissions to manage access

### Any deployment steps required?

> No

### Are any Feature Flags and/or Data Migrations that can eventually be removed Added?

Data Migrations:
-`app-modules/task/database/migrations/2024_12_18_104302_delete_task_individual_update_permissions.php`

_______________________________________________

### CODE REVIEW

File: app-modules/task/src/Policies/TaskPolicy.php [hunk 1]
```diff
@@ -99,10 +99,11 @@ public function update(Authenticatable $authenticatable, Task $task): Response
             return Response::deny('You do not have permission to update this task.');
         }
 
-        return $authenticatable->canOrElse(
-            abilities: ["task.{$task->getKey()}.update"],
-            denyResponse: 'You do not have permission to update this task.'
-        );
```
Comment (Orrison) [inline app-modules/task/src/Policies/TaskPolicy.php:105]: I think we still need what you deleted here after the `assigned_to` and `created_by` check instead of `Response::allow()` because the User should still have to have `test.*.update` which this will cover.

File: app-modules/task/src/Policies/TaskPolicy.php [hunk 2]
```diff
@@ -99,10 +99,11 @@ public function update(Authenticatable $authenticatable, Task $task): Response
...
+        if ($authenticatable->getKey() !== $task?->assigned_to && $authenticatable->getKey() !== $task?->created_by) {
```
Comment (Orrison) [inline app-modules/task/src/Policies/TaskPolicy.php:102]: I don't think you need the `nullsafe` on `$task` in both instances here. There is not a scenario in this method where `$task` can be null

File: app-modules/task/tests/EditTaskTest.php [hunk 3]
```diff
@@ -53,7 +54,9 @@
 test('EditTask is gated with proper access control', function () {
```
Comment (Orrison) [inline app-modules/task/tests/EditTaskTest.php:54]: Because of the nature of the change here this test will need to be modified are changed to ensure that a User must have the `task.view-any` and `task.*.update` permission and either be the creator or currently assigned.

We need to test that all scenarios are covered and the proper response is returned.

File: app-modules/task/tests/TaskAssignmentTest.php [hunk 4]
```diff
@@ -34,49 +34,42 @@
...
-
-it('gives the proper permission to the creator of a Task', function () {
-    /** @var Task $task */
-    $task = Task::factory()->create();
-
-    expect($task->createdBy->can("task.{$task->getKey()}.update"))->toBeTrue();
-});
-
-it('gives the proper permission to the assigned User of a Task on create and update', function () {
+it('gives the proper permission to the assigned User of a Task on update', function () {
```
Comment (Orrison) [inline app-modules/task/tests/TaskAssignmentTest.php:50]: I think we can just remove this test as well. Task Assignment now does not create or do anything to the access of the User directly.

Access is intrinsic based on the assignment. Which can be tested in the Edit page test.

**Review by Orrison — CHANGES_REQUESTED** [review]

File: app-modules/task/tests/EditTaskTest.php [hunk 5]
```diff
@@ -80,12 +92,12 @@
...
+    livewire(EditTask::class, [
         'record' => $task->getRouteKey(),
     ])
         ->fillForm($request->toArray())
         ->call('save')
         ->assertHasNoFormErrors();
 
     // TODO: Check for changes
-});
+})->only();
```
Comment (Orrison) [inline app-modules/task/tests/EditTaskTest.php:103]: This `only` needs to be removed. Please be careful to ensure these are not present before requesting review.

I think all team members should do a code review of their own PRs while in Draft before opening them up for peer review. To check for things like this and to compare against everything listed in the ticket.

File: app-modules/task/tests/EditTaskTest.php
Comment (Orrison) [inline app-modules/task/tests/EditTaskTest.php:1]: The changes you did in this test look good. But let's split this test into multiple to cover all scenarios:

- A user without proper permissions and that is not associated to the Task cannot access
- A User with proper permissions and that is not associated to the Task cannot access
- A User without proper permissions that is the assigned user cannot access
- A User without proper permissions that is the created by user cannot access
- A User with proper permissions that is the assigned user can access
- a User with proper permissions that is the created by user can access.

**Review by Orrison — CHANGES_REQUESTED** [review]

File: app-modules/task/tests/EditTaskTest.php (hunk 5 above)
Comment (payal-canyon) [reply app-modules/task/tests/EditTaskTest.php:103]: Hi @Orrison,
I'm sorry about that. I'll take care of this next time.

Thank You!

Comment (Orrison) [reply app-modules/task/tests/EditTaskTest.php]: No problem!

**Review by Orrison — APPROVED** [review]
